flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/NonTimeRangeUnboundedPrecedingFunction.java

```diff
@@ -0,0 +1,692 @@
+                // Accumulate all ids except id which needs to be removed
+                int removeIndex = reAccumulateIdsAndGetRemoveIndex(ids, delRow);
+                if (removeIndex == -1) {
+                    LOG.info(
+                            "Could not find matching row to remove. Missing id from sortKey ids list.");
+                    numOfIdsNotFound.inc();
+                    return;
+                }
+                emitUpdatesForIds(
+                        ids,
+                        removeIndex,
+                        accMapState.get(curSortKey), // prevAcc
+                        aggFuncs.getAccumulators(), // currAcc
+                        RowKind.DELETE,
+                        delRow,
+                        out);
+                Long deletedId = ids.remove(removeIndex);
+                i = removeIdFromSortedList(sortedList, i, ids, curSortKey);
+                removeFromState(
+                        sortedList, ids.size(), deletedId, curSortKey, aggFuncs.getAccumulators());
+                break;
+            }
+        }
+
+        if (!sortKeyFound) {
+            LOG.debug("Could not find matching sort key. Skipping delete.");
+            numOfSortKeysNotFound.inc();
+        }
+
+        processRemainingElements(sortedList, i, aggFuncs.getAccumulators(), out);
+    }
+
+    /**
+     * Helper method to re-accumulate the aggregated value for all ids without the id that will be
+     * removed.
+     *
+     * @param ids
+     * @param delRow
+     * @return the index position of the id that should be removed
+     * @throws Exception
+     */
+    private int reAccumulateIdsAndGetRemoveIndex(List<Long> ids, RowData delRow) throws Exception {
+        int removeIndex = -1;
+        for (int j = 0; j < ids.size(); j++) {
+            RowData curValue = valueMapState.get(ids.get(j));
+            if (valueEqualiser.equals(curValue, delRow)) {
+                removeIndex = j;
+            } else {
+                aggFuncs.accumulate(curValue);
+            }
+        }
+        return removeIndex;
+    }
+
+    /**
+     * Remove the ith element from the sortedList if the list of ids is empty and return the ith
+     * location OR update the sorted list with the list of ids and return the next location of the
+     * sortedList.
+     *
+     * @param sortedList
+     * @param idx
+     * @param ids
+     * @param curSortKey
+     * @return
+     */
+    private int removeIdFromSortedList(
+            List<Tuple2<RowData, List<Long>>> sortedList,
+            int idx,
+            List<Long> ids,
+            RowData curSortKey) {
+        if (ids.isEmpty()) {
+            sortedList.remove(idx);
+        } else {
+            if (ids.isEmpty()) {
+                sortedList.remove(idx);
+            } else {
+                // Update sortedList without the removed id
+                sortedList.set(idx, new Tuple2<RowData, List<Long>>(curSortKey, ids));
+                idx++;
+            }
```

Comment:
```suggestion
            // Update sortedList without the removed id
            sortedList.set(idx, new Tuple2<RowData, List<Long>>(curSortKey, ids));
            idx++;
```
`ids.isEmpty()` was already checked before.